lib/app/pages/books_screen.dart

```diff
@@ -12,90 +19,340 @@ class BookScreen extends ConsumerStatefulWidget {
-                                  Padding(
-                                    padding: const EdgeInsets.all(8.0),
-                                    child: Text(
-                                      book.displayTitle ?? "nuull",
-                                      style: const TextStyle(
-                                        fontWeight: FontWeight.bold,
-                                        color: Colors.black,
-                                        fontSize: 24,
+                                  if (chapter.chapters.isNotEmpty)
+                                    ExpansionTile(
+                                      title: Text(
+                                        chapter.bookTitle,
+                                        style: const TextStyle(fontWeight: FontWeight.normal, fontSize: 20),
                                       ),
-                                    ),
+                                      children: chapter.chapters
+                                          .map(
+                                            (chapter) => InkWell(
+                                              onTap: () {
+                                                if (chapter.valid) {
+                                                  setState(() {
+                                                    markChapterAsRead(chapter.id.toString());
+                                                  });
+                                                }
+                                              },
+                                              child: ListTile(
+                                                leading: SizedBox(
+                                                  width: 50,
+                                                  child: CachedNetworkImage(
+                                                    imageUrl: chapter.url,
+                                                    height: 140,
+                                                    width: double.infinity,
+                                                    fit: BoxFit.cover,
+                                                    errorWidget: (context, url, error) =>
+                                                        Image.asset("assets/img/placeholder.png"),
+                                                  ),
+                                                ),
+                                                title: Text(
+                                                  chapter.title,
+                                                  style: TextStyle(
+                                                    color: Paint().color = chapter.valid ? Colors.black : Colors.grey,
+                                                  ),
+                                                ),
+                                                trailing: FutureBuilder<bool>(
+                                                  future: isChapterRead(chapter.id.toString()),
```

Comment:
This creates an async component for each chapter; isn't there a more efficient method, like loading the list of read chapters once, and looking up read chapters from that list rather than the shared preferences?